src/R/Core/Impl/AST/Extensions/ScopeExtensions.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.Common.Core;
+using Microsoft.R.Core.AST.Arguments;
+using Microsoft.R.Core.AST.DataTypes;
+using Microsoft.R.Core.AST.Functions.Definitions;
+using Microsoft.R.Core.AST.Scopes;
+using Microsoft.R.Core.AST.Scopes.Definitions;
+using Microsoft.R.Core.AST.Statements.Definitions;
+using Microsoft.R.Core.AST.Variables;
+
+namespace Microsoft.R.Core.AST {
+    public static class ScopeExtensions {
+        /// <summary>
+        /// Enumerates scopes from the current up to the global scope.
+        /// </summary>
+        public static IEnumerable<IScope> EnumerateTowardsGlobal(this IScope scope) {
+            yield return scope;
+            while (!(scope is GlobalScope)) {
+                scope = scope.GetEnclosingScope();
+                yield return scope;
+            }
+        }
+
+        /// <summary>
+        /// Enumerates definitions of variables applicable to the given scope.
+        /// Traverses scopes from the provided scope up and enumerates
+        /// variables that appear at the left side of the left-hand assignment
+        /// operator or at the right side of the right-hand assignmend operator.
+        /// Includes both regular variables as well as function definitions.
+        /// Includes assignments that appear in the file up to the specified
+        /// position except in the global scope it enumerates all assignments.
```

Comment:
It is an interesting question... Originally I did it line this. However, the behavior then changes depending on if file was sourced or not. If it was, then all the functions in global scope become available in the editor. Same happens in RS. So current behavior is identical before and after sourcing file. Also, I have files from R itself (like check.r) where code in the middle of the file references functions declared in the end.